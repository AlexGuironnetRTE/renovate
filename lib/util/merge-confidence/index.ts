import type { UpdateType } from '../../config/types';
import { logger } from '../../logger';
import * as hostRules from '../host-rules';
import { Http } from '../http';
import * as cache from './cache';

const http = new Http('merge-confidence');

const MERGE_CONFIDENCE = ['low', 'neutral', 'high', 'very high'];
type MergeConfidenceTuple = typeof MERGE_CONFIDENCE;
export type MergeConfidence = MergeConfidenceTuple[number];

export const confidenceLevels: Record<MergeConfidence, number> = {
  low: -1,
  neutral: 0,
  high: 1,
  'very high': 2,
};

export function isActiveConfidenceLevel(confidence: string): boolean {
  return confidence !== 'low' && MERGE_CONFIDENCE.includes(confidence);
}

export function satisfiesConfidenceLevel(
  confidence: MergeConfidence,
  minimumConfidence: MergeConfidence
): boolean {
  return confidenceLevels[confidence] >= confidenceLevels[minimumConfidence];
}

const updateTypeConfidenceMapping: Record<UpdateType, MergeConfidence> = {
  pin: 'high',
  digest: 'neutral',
  bump: 'neutral',
  lockFileMaintenance: 'neutral',
  lockfileUpdate: 'neutral',
  rollback: 'neutral',
  major: null,
  minor: null,
  patch: null,
};

export async function getMergeConfidenceLevel(
  datasource: string,
  depName: string,
  currentVersion: string,
  newVersion: string,
  updateType: UpdateType
): Promise<MergeConfidence> {
  if (!(currentVersion && newVersion && updateType)) {
    return 'neutral';
  }
  const mappedConfidence = updateTypeConfidenceMapping[updateType];
  if (mappedConfidence) {
    return mappedConfidence;
  }
  const { token } = hostRules.find({
    hostType: 'merge-confidence',
    url: 'https://badges.renovateapi.com',
  });
  if (!token) {
    logger.warn('No Merge Confidence API token found');
    return 'neutral';
  }
  const url = `https://badges.renovateapi.com/packages/${datasource}/${depName}/${newVersion}/confidence.api/${currentVersion}`;
  const cachedResult = cache.get(token, url);
  if (cachedResult) {
    return cachedResult;
  }
  let confidence = 'neutral';
  try {
    const res = (await http.getJson<{ confidence: MergeConfidence }>(url)).body;
    if (MERGE_CONFIDENCE.includes(res.confidence)) {
      confidence = res.confidence;
    }
  } catch (err) {
    logger.debug({ err }, 'Error fetching merge confidence');
    if (err.statusCode === 403) {
      logger.warn('Merge Confidence API token rejected');
    }
  }
  cache.set(token, url, confidence);
  return confidence;
}
