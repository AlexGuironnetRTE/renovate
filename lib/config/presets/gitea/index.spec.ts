import * as httpMock from '../../../../test/http-mock';
import { getName, mocked } from '../../../../test/util';
import * as _hostRules from '../../../util/host-rules';
import { setBaseUrl } from '../../../util/http/gitea';
import { PRESET_INVALID_JSON, PRESET_NOT_FOUND } from '../util';
import * as gitea from '.';

jest.mock('../../../util/host-rules');

const hostRules = mocked(_hostRules);

const giteaApiHost = gitea.Endpoint;
const basePath = '/repos/some/repo/contents';

describe(getName(), () => {
  beforeEach(() => {
    hostRules.find.mockReturnValue({ token: 'abc' });
    setBaseUrl(giteaApiHost);
  });

  describe('fetchJSONFile()', () => {
    it('returns JSON', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/some-filename.json`)
        .reply(200, {
          content: Buffer.from('{"from":"api"}').toString('base64'),
        });

      const res = await gitea.fetchJSONFile(
        'some/repo',
        'some-filename.json',
        giteaApiHost
      );
      expect(res).toMatchSnapshot();
    });
  });

  describe('getPreset()', () => {
    it('tries default then renovate', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/default.json`)
        .reply(404, {})
        .get(`${basePath}/renovate.json`)
        .reply(200, {});

      await expect(
        gitea.getPreset({ packageName: 'some/repo' })
      ).rejects.toThrow();
    });

    it('throws if no content', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/default.json`)
        .reply(200, {});

      await expect(
        gitea.getPreset({ packageName: 'some/repo' })
      ).rejects.toThrow(PRESET_INVALID_JSON);
    });

    it('throws if fails to parse', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/default.json`)
        .reply(200, {
          content: Buffer.from('not json').toString('base64'),
        });

      await expect(
        gitea.getPreset({ packageName: 'some/repo' })
      ).rejects.toThrow(PRESET_INVALID_JSON);
    });

    it('should return default.json', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/default.json`)
        .reply(200, {
          content: Buffer.from('{"foo":"bar"}').toString('base64'),
        });

      const content = await gitea.getPreset({ packageName: 'some/repo' });
      expect(content).toEqual({ foo: 'bar' });
    });

    it('should query preset within the file', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/somefile.json`)
        .reply(200, {
          content: Buffer.from('{"somename":{"foo":"bar"}}').toString('base64'),
        });
      const content = await gitea.getPreset({
        packageName: 'some/repo',
        presetName: 'somefile/somename',
      });
      expect(content).toEqual({ foo: 'bar' });
    });

    it('should query subpreset', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/somefile.json`)
        .reply(200, {
          content: Buffer.from(
            '{"somename":{"somesubname":{"foo":"bar"}}}'
          ).toString('base64'),
        });

      const content = await gitea.getPreset({
        packageName: 'some/repo',
        presetName: 'somefile/somename/somesubname',
      });
      expect(content).toEqual({ foo: 'bar' });
    });

    it('should return custom.json', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/custom.json`)
        .reply(200, {
          content: Buffer.from('{"foo":"bar"}').toString('base64'),
        });
      const content = await gitea.getPreset({
        packageName: 'some/repo',
        presetName: 'custom',
      });
      expect(content).toEqual({ foo: 'bar' });
    });

    it('should query custom paths', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/path%2Fcustom.json`)
        .reply(200, {
          content: Buffer.from('{"foo":"bar"}').toString('base64'),
        });
      const content = await gitea.getPreset({
        packageName: 'some/repo',
        presetName: 'custom',
        presetPath: 'path',
      });
      expect(content).toEqual({ foo: 'bar' });
    });

    it('should throws not-found', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/somefile.json`)
        .reply(200, {
          content: Buffer.from('{}').toString('base64'),
        });
      await expect(
        gitea.getPreset({
          packageName: 'some/repo',
          presetName: 'somefile/somename/somesubname',
        })
      ).rejects.toThrow(PRESET_NOT_FOUND);
    });
  });

  describe('getPresetFromEndpoint()', () => {
    it('uses default endpoint', async () => {
      httpMock
        .scope(giteaApiHost)
        .get(`${basePath}/default.json`)
        .reply(200, {
          content: Buffer.from('{"from":"api"}').toString('base64'),
        });
      expect(
        await gitea.getPresetFromEndpoint('some/repo', 'default', undefined)
      ).toEqual({ from: 'api' });
    });

    it('uses custom endpoint', async () => {
      httpMock
        .scope('https://api.gitea.example.org')
        .get(`${basePath}/default.json`)
        .reply(200, {
          content: Buffer.from('{"from":"api"}').toString('base64'),
        });
      expect(
        await gitea
          .getPresetFromEndpoint(
            'some/repo',
            'default',
            undefined,
            'https://api.gitea.example.org'
          )
          .catch(() => ({ from: 'api' }))
      ).toEqual({ from: 'api' });
    });
  });
});
