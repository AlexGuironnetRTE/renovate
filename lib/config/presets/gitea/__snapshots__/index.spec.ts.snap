// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config/presets/gitea/index fetchJSONFile() returns JSON 1`] = `
Object {
  "from": "api",
}
`;
